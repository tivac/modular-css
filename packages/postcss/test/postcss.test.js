"use strict";

const fs = require("fs");
const postcss = require("postcss");
const read    = require("@modular-css/test-utils/read.js")(__dirname);
const namer   = require("@modular-css/test-utils/namer.js");
const plugin  = require("../postcss.js");

function process(file, opts = {}) {
    return plugin.process(
        fs.readFileSync(file),
        {
            __proto__ : null,

            from : file,
            namer,
            ...opts,
        }
    );
}

describe("/postcss.js", () => {
    afterAll(() => require("shelljs").rm("-rf", "./packages/postcss/test/output/*"));
    
    it("should be a function", () => {
        expect(typeof plugin).toBe("function");
    });

    it("should process CSS and output the result", async () => {
        const { css } = await process("./packages/postcss/test/specimens/simple.css");
        
        expect(css).toMatchSnapshot();
    });

    it("should process CSS with dependencies and output the result", async () => {
        const { css } = await process("./packages/postcss/test/specimens/start.css");
        
        expect(css).toMatchSnapshot();
    });

    it("should process CSS and output exports as a message", async () => {
        const { messages } = await process("./packages/postcss/test/specimens/simple.css");
        
        expect(messages).toMatchSnapshot();
    });

    it("should accept normal processor options", async () => {
        const { css } = await process("./packages/postcss/test/specimens/simple.css", {
            map : {
                inline : true,
            },
            namer : (f, s) => `fooga_${s}`,
        });
    
        expect(css).toMatchSnapshot();
    });

    it("should accept a `json` property and write exports to that file", async () => {
        await process(
            "./packages/postcss/test/specimens/start.css",
            {
                json : "./packages/postcss/test/output/classes.json",
            }
        );
        
        expect(read("classes.json")).toMatchSnapshot();
    });

    it("should use output filepath for json if a custom path isn't provided", async () => {
        await process(
            "./packages/postcss/test/specimens/start.css",
            {
                json : true,
                to   : "./packages/postcss/test/output/start.css",
            }
        );
        
        expect(read("start.json")).toMatchSnapshot();
    });

    it("should be usable like a normal postcss plugin", async () => {
        const processor = postcss([
            plugin({
                namer : () => "a",
            }),
        ]);
        
        const { css } = await processor.process(
            fs.readFileSync("./packages/postcss/test/specimens/simple.css"),
            {
                from : "./packages/postcss/test/specimens/simple.css",
                map  : {
                    inline : true,
                },
            }
        );
    
        expect(css).toMatchSnapshot();
    });

    it("should output json when used within postcss", async () => {
        const processor = postcss([
            plugin({
                namer,
            }),
        ]);
        
        await processor.process(
            fs.readFileSync("./packages/postcss/test/specimens/simple.css"),
            {
                from : "./packages/postcss/test/specimens/simple.css",
                json : "./packages/postcss/test/output/simple.json",
            }
        );
        
        expect(read("simple.json")).toMatchSnapshot();
    });

    it("should accept json args in either position with postcss", async () => {
        const processor = postcss([
            plugin({
                namer,
                json : "./packages/postcss/test/output/simple.json",
            }),
        ]);
        
        await processor.process(
            fs.readFileSync("./packages/postcss/test/specimens/simple.css"),
            {
                from : "./packages/postcss/test/specimens/simple.css",
            }
        );

        expect(read("simple.json")).toMatchSnapshot();
    });
});
