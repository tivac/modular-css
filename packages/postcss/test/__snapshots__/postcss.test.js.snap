// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/postcss.js should accept a \`json\` property and write exports to that file 1`] = `
"{
    \\"packages/postcss/test/specimens/folder/folder.css\\": {
        \\"folder\\": \\"folder\\"
    },
    \\"packages/postcss/test/specimens/local.css\\": {
        \\"booga\\": \\"booga\\",
        \\"looga\\": \\"booga looga\\"
    },
    \\"packages/postcss/test/specimens/start.css\\": {
        \\"wooga\\": \\"booga wooga\\",
        \\"booga\\": \\"booga\\",
        \\"tooga\\": \\"tooga\\"
    }
}"
`;

exports[`/postcss.js should accept json args in either position with postcss 1`] = `
"{
    \\"packages/postcss/test/specimens/simple.css\\": {
        \\"wooga\\": \\"wooga\\"
    }
}"
`;

exports[`/postcss.js should accept normal processor options 1`] = `
".fooga_wooga { color: red; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL3Bvc3Rjc3MvdGVzdC9zcGVjaW1lbnMvc2ltcGxlLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxlQUFTLFVBQVUsRUFBRSIsImZpbGUiOiJwYWNrYWdlcy9wb3N0Y3NzL3Rlc3Qvc3BlY2ltZW5zL3NpbXBsZS5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyIud29vZ2EgeyBjb2xvcjogcmVkOyB9XG4iXX0= */"
`;

exports[`/postcss.js should be usable like a normal postcss plugin 1`] = `
".a { color: red; }

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInBhY2thZ2VzL3Bvc3Rjc3MvdGVzdC9zcGVjaW1lbnMvc2ltcGxlLmNzcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQSxLQUFTLFVBQVUsRUFBRSIsImZpbGUiOiJwYWNrYWdlcy9wb3N0Y3NzL3Rlc3Qvc3BlY2ltZW5zL3NpbXBsZS5jc3MiLCJzb3VyY2VzQ29udGVudCI6WyIud29vZ2EgeyBjb2xvcjogcmVkOyB9XG4iXX0= */"
`;

exports[`/postcss.js should output json when used within postcss 1`] = `
"{
    \\"packages/postcss/test/specimens/simple.css\\": {
        \\"wooga\\": \\"wooga\\"
    }
}"
`;

exports[`/postcss.js should process CSS and output exports as a message 1`] = `
Array [
  Object {
    "exports": Object {
      "wooga": Array [
        "wooga",
      ],
    },
    "type": "modular-css-exports",
  },
]
`;

exports[`/postcss.js should process CSS and output the result 1`] = `
".wooga { color: red; }
"
`;

exports[`/postcss.js should process CSS with dependencies and output the result 1`] = `
".booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/postcss.js should use output filepath for json if a custom path isn't provided 1`] = `
"{
    \\"packages/postcss/test/specimens/folder/folder.css\\": {
        \\"folder\\": \\"folder\\"
    },
    \\"packages/postcss/test/specimens/local.css\\": {
        \\"booga\\": \\"booga\\",
        \\"looga\\": \\"booga looga\\"
    },
    \\"packages/postcss/test/specimens/start.css\\": {
        \\"wooga\\": \\"booga wooga\\",
        \\"booga\\": \\"booga\\",
        \\"tooga\\": \\"tooga\\"
    }
}"
`;
